src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -1018,7 +1033,76 @@ private void EmitProviders(ExpressionEmitter emitter, ImmutableArray<DeclaredPro
             });
         }
 
-        private DeclaredProviderExpression GetProviderForLocalDeploy()
+        private void EmitExtensions(ExpressionEmitter emitter, ImmutableArray<DeclaredProviderExpression> providers)
+        {
+            emitter.EmitObjectProperty("extensions", () =>
+            {
+                foreach (var provider in providers)
+                {
+                    var settings = provider.Settings;
+
+                    emitter.EmitObjectProperty(provider.Name, () =>
+                    {
+                        emitter.EmitProperty("name", settings.ArmTemplateProviderName);
+                        emitter.EmitProperty("version", settings.ArmTemplateProviderVersion);
+
+                        EmitExtensionConfig(provider, emitter);
+                    },
+                    provider.SourceSyntax);
+                }
+            });
+        }
+
+        private void EmitExtensionConfig(DeclaredProviderExpression provider, ExpressionEmitter emitter)
+        {
+            if (provider.Config is null)
+            {
+                return;
+            }
+
+            if (provider.Config is not ObjectExpression providerConfig)
+            {
+                throw new UnreachableException($"Provider config type expected to be of type: '{nameof(ObjectExpression)}' but received: '{provider.Config.GetType()}'");
+            }
+
+            emitter.EmitObjectProperty("config", () =>
+            {
+                foreach (var providerConfigProperty in providerConfig.Properties)
+                {
+                    // Type checking should have validated that the config name is not an expression (e.g. string interpolation), if we get a null value it means something
+                    // was wrong with type checking validation.
+                    var extensionConfigName = providerConfigProperty.TryGetKeyText() ?? throw new UnreachableException("Expressions are not allowed as config names.");
+                    var configType = provider.Settings.ConfigurationType ?? throw new UnreachableException("Config type must be specified.");
+                    var extensionConfigType = GetExtensionConfigType(extensionConfigName, configType);
+
+                    emitter.EmitObjectProperty(extensionConfigName, () =>
+                    {
+                        switch (extensionConfigType)
+                        {
+                            case StringType:
+                                emitter.EmitProperty("type", "string");
+                                break;
```

Comment:
Is it worth handling `secureString` here?

Reply:
I am avoiding it because `secureString` requires bicep to read the value from KeyVault and not sure if all the wiring is in place right now.

Reply:
after chatting with Shenglong, will be including the same types that we support today in Deployments